### PR TITLE
[Unity][Transform] Implement relax.transform.AdjustMatmulOrder

### PR DESCRIPTION
**Pull request by Lunderberg**
Reorder `x*(A*B)` to `(x*A)*B`.  Intended for optimization of LoRA models, for which `(x*A)*B` has a much smaller memory footprint.

### CODE REVIEW

**Comment by Lunderberg** [comment]
This PR is currently a draft, as it depends on functionality introduced in https://github.com/apache/tvm/pull/16312.

**Comment by Lunderberg** [comment]
The prerequisite #16312 has now landed, and this PR is ready for review.

File: src/relax/transform/adjust_matmul_order.cc [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tvm/relax/transform/expand_matmul_of_sum.cc
+ * \brief Expand `matmul(x, A+B)` to `matmul(x, A) + matmul(x,B)`
+ */
```
Comment (Hzfengsy) [inline src/relax/transform/adjust_matmul_order.cc:23]: typo :)

**Review by Hzfengsy — COMMENTED** [review]
Thanks for the PR, which would help on the specific case. However, I think it's necessary to analyze the memory footprint before rewriting it. 

it would be good to rewrite `x * (A * B) -> (x * A) * B`, but not reasonable to rewrite `A * (B * x) -> (A * B) * x

File: src/relax/transform/adjust_matmul_order.cc (hunk 1 above)
Comment (Lunderberg) [reply src/relax/transform/adjust_matmul_order.cc]: Thank you, and fixed.

**Comment by Lunderberg** [comment]
Good point.  I've updated the pattern and the rewrite rule with two changes:

1. Check if `(A*B)*C` should be re-ordered into `A*(B*C)`.  Previously, only the reverse was checked.
2. Only reorder the matmuls in either case when there's a provable benefit to doing so, based on the shapes of the three matrices.

I also added several additional unit tests to validate the behavior, for cases where the initial matmul is left-to-right, and where the initial matmul is right-to-left.

**Comment by Lunderberg** [comment]
@Hzfengsy All changes based on your earlier comments have been made, and the PR is ready for re-review.

**Review by Hzfengsy — APPROVED** [review]
LGTM. Thanks for the great work!

**Pull request merged**